corehq/ex-submodules/dimagi/utils/django/email.py

```diff
@@ -17,10 +17,30 @@
 LARGE_FILE_SIZE_ERROR_CODES = [LARGE_FILE_SIZE_ERROR_CODE, LARGE_FILE_SIZE_ERROR_CODE_ICDS_TCL]
 
 
+def get_valid_recipients(recipients):
+    """
+    This filters out any emails that have reported hard bounces or complaints to
+    Amazon SES
+    :param recipients: list of recipient emails
+    :return: list of recipient emails not marked as bounced
+    """
+    from corehq.util.models import BouncedEmail
+    valid_recipients = []
+    for recipient in recipients:
+        if not BouncedEmail.objects.filter(email=recipient).exists():
+            valid_recipients.append(recipient)
+    return valid_recipients
```

Comment:
Could replace this line and the for loop with something like

```
bounced_emails = set(BouncedEmail.objects.filter(email__in=recipients).values_list('email', flat=True))
return [recipient for recipient in recipients if recipient not in bounced_emails]
```
which would do the same thing in fewer queries. I suppose if there are ever thousands of items in `recipients`, we could iterate over it in chunks of 500 or something (to avoid the query text itself being too large), not sure if that's a consideration.

Reply:
yes, definitely a good idea. I will make this change in followup PR.